temporal-sdk/src/test/java/io/temporal/internal/statemachines/UpdateProtocolStateMachineTest.java

```diff
@@ -396,4 +400,486 @@ protected void update(UpdateMessage message, AsyncWorkflowBuilder<Void> builder)
+                  message.getCallbacks().accept();
+                })
+            .add(
+                (r) -> {
+                  message.getCallbacks().complete(converter.toPayloads("update result"), null);
+                })
+            .add((r) -> stateMachines.completeWorkflow(Optional.empty()));
+      }
+    }
+
+    /*
+        1: EVENT_TYPE_WORKFLOW_EXECUTION_STARTED
+        2: EVENT_TYPE_WORKFLOW_TASK_SCHEDULED
+        3: EVENT_TYPE_WORKFLOW_TASK_STARTED
+        4: EVENT_TYPE_WORKFLOW_TASK_COMPLETED
+        5: EVENT_TYPE_WORKFLOW_EXECUTION_UPDATE_ADMITTED
+        6: EVENT_TYPE_WORKFLOW_TASK_SCHEDULED
+        7: EVENT_TYPE_WORKFLOW_TASK_STARTED
+        8: EVENT_TYPE_WORKFLOW_TASK_COMPLETED
+        9: EVENT_TYPE_WORKFLOW_EXECUTION_UPDATE_ACCEPTED
+        10: EVENT_TYPE_WORKFLOW_EXECUTION_UPDATE_COMPLETED
+        11: EVENT_TYPE_WORKFLOW_EXECUTION_COMPLETED
+    */
+
+    TestHistoryBuilder h = new TestHistoryBuilder();
+    {
+      TestEntityManagerListenerBase listener = new TestUpdateListener();
+      stateMachines = newStateMachines(listener);
+      h.add(EventType.EVENT_TYPE_WORKFLOW_EXECUTION_STARTED);
+      h.addWorkflowTask();
+      h.add(
+          EventType.EVENT_TYPE_WORKFLOW_EXECUTION_UPDATE_ADMITTED,
+          WorkflowExecutionUpdateAdmittedEventAttributes.newBuilder()
+              .setRequest(
+                  Request.newBuilder()
+                      .setMeta(Meta.newBuilder().setUpdateId("protocol_id").build())
+                      .setInput(
+                          Input.newBuilder()
+                              .setName("updateName")
+                              .setArgs(converter.toPayloads("arg").get()))));
+      h.addWorkflowTask();
+      h.add(
+          EventType.EVENT_TYPE_WORKFLOW_EXECUTION_UPDATE_ACCEPTED,
+          WorkflowExecutionUpdateAcceptedEventAttributes.newBuilder()
+              .setProtocolInstanceId("protocol_id")
+              .setAcceptedRequestMessageId("id")
+              .setAcceptedRequestSequencingEventId(2));
+      h.add(
+          EventType.EVENT_TYPE_WORKFLOW_EXECUTION_UPDATE_COMPLETED,
+          WorkflowExecutionUpdateCompletedEventAttributes.newBuilder()
+              .setOutcome(
+                  Outcome.newBuilder().setSuccess(converter.toPayloads("m2Arg1").get()).build()));
+      h.add(EventType.EVENT_TYPE_WORKFLOW_EXECUTION_COMPLETED);
+    }
+    {
+      TestEntityManagerListenerBase listener = new TestUpdateListener();
+      stateMachines = newStateMachines(listener);
+      List<Command> commands = h.handleWorkflowTaskTakeCommands(stateMachines, 2);
+      assertEquals(3, commands.size());
+    }
+    {
+      TestEntityManagerListenerBase listener = new TestUpdateListener();
+      stateMachines = newStateMachines(listener);
+      List<Command> commands = h.handleWorkflowTaskTakeCommands(stateMachines);
+      assertEquals(0, commands.size());
+    }
+  }
+
+  @Test
+  public void testUpdateSignalSandwichAdmittedMessage() {
```

Comment:
@Sushisource @dandavison This is something we should test in Core as well, previously due to server limitations, it wasn't possible to have an `[update, signal, update]` in history, but with update admitted it now is. We should check the handlers get called in the correct order.